src/jit/lowerxarch.cpp

```diff
@@ -921,11 +921,17 @@ void Lowering::TreeNodeInfoInit(GenTree* tree)
     //         cmp dl, byte ptr[addr mode]
     //         movzx edx, dl
     //
+    // Example4: GT_EQ(int, op1 of type ubyte, op2 is GT_CNS_INT) - in this case codegen uses
+    // ubyte as the result of the comparison and if the result needs to be materialized into a reg
+    // simply zero extend it to TYP_INT size.
+    //
     // Though this looks conservative in theory, in practice we could not think of a case where
     // the below logic leads to conservative register specification.  In future when or if we find
     // one such case, this logic needs to be fine tuned for that case(s).
     if (varTypeIsByte(tree) || ((tree->OperGet() == GT_CAST) && varTypeIsByte(tree->CastToType())) ||
-        (tree->OperIsCompare() && varTypeIsByte(tree->gtGetOp1()) && varTypeIsByte(tree->gtGetOp2())))
+        (tree->OperIsCompare() && ((varTypeIsByte(tree->gtGetOp1()) &&
+         (varTypeIsByte(tree->gtGetOp2()) || tree->gtGetOp2()->IsCnsIntOrI())) ||
+         (varTypeIsByte(tree->gtGetOp2()) && tree->gtGetOp1()->IsCnsIntOrI()))))
     {
```

Comment:
We have #7527 opened to restructure this code, so this won't stay around for very long.